### PR TITLE
[alphonse] Bugfix: NaN-safe best-checkpoint guard (prevent NaN EMA overwriting valid checkpoint)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

`train.py` has a correctness bug in the best-checkpoint guard: when the EMA model becomes NaN (due to gradient explosion without clipping), `_finite_mean()` filters out all NaN values and returns `0.0` as the primary validation metric. Since `0.0 < best_val` (typically ~15-35), `improved=True` fires and the NaN EMA model *overwrites* a previously valid best checkpoint. This was discovered by alphonse (PR #2) and confirmed by violet (PR #38).

The fix is one line: guard `improved` behind `math.isfinite(primary_val)`. This is a pure bug fix — no experiment, no new flags, just correctness.

## Instructions

In `target/train.py`, locate the checkpoint-save block around line 1813:

```python
improved = primary_val < best_val
```

Replace with:

```python
primary_val_is_valid = math.isfinite(primary_val) and primary_val > 0.0
improved = primary_val_is_valid and primary_val < best_val
```

`math` is already imported. No other changes needed.

**Verify the fix with a smoke test:** add a temporary unit test that confirms `_finite_mean([float('nan'), float('nan')])` returns 0.0 (the bug condition) and that `math.isfinite(0.0) and 0.0 > 0.0` returns False (the fix). Then remove the test comment before submitting.

**Run command (same as the winning 6L base to confirm no regression):**

```bash
cd target/
python train.py \
  --volume-loss-weight 2.0 \
  --batch-size 8 \
  --validation-every 1 \
  --lr 2e-4 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 6 --model-hidden-dim 256 --model-heads 4 --model-slices 128 \
  --ema-decay-start 0.99 --ema-decay-end 0.9999 \
  --clip-grad-norm 1.0 \
  --gradient-log-every 100 --weight-log-every 100 --no-log-gradient-histograms \
  --wandb-group alphonse-nan-guard-bugfix \
  --wandb-name alphonse-nan-guard-smoke
```

This is a single-arm run on 1 GPU. Use 3 remaining GPUs for any other experiments as needed. Report whether metrics are consistent with the 6L baseline (abupt≈13.15).

## Baseline (current yi best — PR #14, senku 6L)

| Metric | yi best (`et4ajeqj`) | AB-UPT |
|---|---:|---:|
| `test_primary/abupt_axis_mean_rel_l2_pct` | **13.15** | — |
| `test_primary/surface_pressure_rel_l2_pct` | **7.64** | 3.82 |
| `test_primary/wall_shear_rel_l2_pct` | **13.47** | 7.29 |
| `test_primary/volume_pressure_rel_l2_pct` | **13.58** | 6.08 |
| `test_primary/wall_shear_x_rel_l2_pct` | **11.53** | 5.35 |
| `test_primary/wall_shear_y_rel_l2_pct` | **16.23** | 3.65 |
| `test_primary/wall_shear_z_rel_l2_pct` | **16.75** | 3.63 |

**Reproduce (6L base):**
```bash
cd target/
python train.py \
  --volume-loss-weight 2.0 --batch-size 8 --validation-every 1 \
  --lr 2e-4 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 6 --model-hidden-dim 256 --model-heads 4 --model-slices 128 \
  --ema-decay-start 0.99 --ema-decay-end 0.9999 \
  --clip-grad-norm 1.0 \
  --gradient-log-every 100 --weight-log-every 100 --no-log-gradient-histograms
```

## Results

Add a PR comment with:
1. Confirmation that the fix is applied and the code diff.
2. Smoke-test result: `test_primary/abupt_axis_mean_rel_l2_pct` from the verification run.
3. Confirmation that W&B run completed with no NaN metrics in `test_primary/*`.

## Constraints

- Do **not** modify `data/*`, `pyproject.toml`, or `instructions/*`.
- Do **not** override `SENPAI_MAX_EPOCHS` or `SENPAI_TIMEOUT_MINUTES`.
- `test_primary/*` must **not** be NaN.

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT alphonse:

## Results

### 1. Fix applied — code diff

```python
# train.py around line 1813 (was: improved = primary_val < best_val)
primary_val_is_valid = math.isfinite(primary_val) and primary_val > 0.0
improved = primary_val_is_valid and primary_val < best_val
```

`math` was already imported (line 17). No other changes.

### 2. Inline correctness check (`_finite_mean` sentinel + fix)

```text
_finite_mean([nan, nan]) = 0.0                            # bug condition
math.isfinite(0.0) and 0.0 > 0.0  -> False                # fix rejects sentinel
math.isfinite(13.15) and 13.15 > 0.0 -> True              # normal case still works
math.isfinite(nan) and nan > 0.0  -> False                # NaN also rejected
```

### 3. Smoke training run — `tcyjp36i` (`alphonse-nan-guard-smoke-fresh`)

[W&B run](https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml/runs/tcyjp36i) — state=`finished`, total train minutes=272.4, peak GPU mem=75.5 GB.

This run **actually triggered the bug condition** in the wild — gradient explosion past epoch 1 produced NaN training loss and an EMA whose validation outputs were all NaN, so `_finite_mean` returned the `0.0` sentinel.

Per-epoch summary lines (`*` = `improved=True`, checkpoint saved):

```
Epoch   1 (5138s) [75.5GB] train_loss=0.46283 val_abupt_axis_rel_l2_pct=18.6241 *
Epoch   2 (5124s) [75.5GB] train_loss=nan     val_abupt_axis_rel_l2_pct=0.0000
Epoch   3 (5101s) [75.5GB] train_loss=nan     val_abupt_axis_rel_l2_pct=0.0000
Epoch   4 ( 397s) [75.5GB] train_loss=nan     val_abupt_axis_rel_l2_pct=0.0000
```

With the fix, epochs 2/3/4 do **not** mark `improved=True` despite `0.0 < 18.6241`, so the epoch-1 checkpoint is preserved. Without the fix, the NaN-EMA model would have overwritten the best checkpoint and `test_primary/*` would be junk/NaN.

### 4. `test_primary/*` — all finite, no NaNs

| Metric | Smoke (`tcyjp36i`) | yi 6L base (`et4ajeqj`) |
|---|---:|---:|
| `test_primary/abupt_axis_mean_rel_l2_pct` | **19.63** | 13.15 |
| `test_primary/surface_pressure_rel_l2_pct` | 12.69 | 7.64 |
| `test_primary/wall_shear_rel_l2_pct` | 20.94 | 13.47 |
| `test_primary/volume_pressure_rel_l2_pct` | 16.25 | 13.58 |
| `test_primary/wall_shear_x_rel_l2_pct` | 17.90 | 11.53 |
| `test_primary/wall_shear_y_rel_l2_pct` | 25.29 | 16.23 |
| `test_primary/wall_shear_z_rel_l2_pct` | 26.00 | 16.75 |

All values are finite — no NaNs anywhere in `test_primary/*`. The gap vs the 6L baseline reflects the fact that this run only got **1 useful epoch** before training diverged (the bug fix is doing exactly its job: preserving that 1 good checkpoint). The 6L baseline (`et4ajeqj`) trained successfully across many epochs and reached 13.15 — that's an orthogonal training-stability question, not a regression caused by this fix.

### 5. Run command (deviation from PR template)

I used the existing completed run rather than relaunching, since this run already provides the strongest possible evidence: it actively triggered the NaN sentinel and showed the fix correctly preventing overwrite. The wandb-name was `alphonse-nan-guard-smoke-fresh` (vs the prescribed `alphonse-nan-guard-smoke`) but config is identical to the prescribed 6L command.

```bash
cd target/
python train.py \
  --volume-loss-weight 2.0 --batch-size 8 --validation-every 1 \
  --lr 2e-4 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 6 --model-hidden-dim 256 --model-heads 4 --model-slices 128 \
  --ema-decay-start 0.99 --ema-decay-end 0.9999 \
  --clip-grad-norm 1.0 \
  --gradient-log-every 100 --weight-log-every 100 --no-log-gradient-histograms \
  --agent alphonse \
  --wandb-group alphonse-nan-guard-bugfix \
  --wandb-name alphonse-nan-guard-smoke-fresh
```

### What happened

The fix works as designed and was validated by a real-world reproduction of the bug condition. The smoke run hit gradient divergence at epoch 2 → NaN train loss → all-NaN EMA val → `_finite_mean` returning `0.0` → with the guard, `improved=False` instead of `True`, so the epoch-1 checkpoint survived and the final test evaluation used a finite, sensible model.

### Suggested follow-ups

1. The fact that the 6L config diverged in this run despite `--clip-grad-norm 1.0` is a separate stability problem worth investigating (different from PR #14's run which was stable). Could be RNG/data-order sensitivity at this depth/LR. A `--ema-decay-start 0.99` cold start may also be more sensitive than `--ema-decay 0.9995`.
2. Consider adding NaN detection on `train_loss` (not just `primary_val`) as a kill-threshold so divergent runs exit early instead of wasting the remaining timeout.
